---
title: "Introduction to Differential Gene Expression Analysis with Bulk RNAseq"
subtitle: "Data exploration"
date: "September 2023"
output:
  ioslides_presentation:
    css: css/stylesheet.css
    logo: images/CRUK_Cambridge_Institute.png
    smaller: yes
    widescreen: yes
  beamer_presentation: default
---

## Introduction

In this session we will:

* import our counts into R  
* look at the effects of variance and how to mitigate this with data 
transformation
* do some initial exploration of the raw count data using principle component 
analysis  


## Sample metadata

First we load a table of sample metadata

```{r loadSampleInfo, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
library(DESeq2)
library(tidyverse)
sampleinfo <- read_tsv("data/samplesheet.tsv", col_types = c("cccc"))
```

```{r loadSampleInfoA, eval=FALSE}
library(tidyverse)
sampleinfo <- read_tsv("data/samplesheet.tsv", col_types = c("cccc"))
arrange(sampleinfo, Status, TimePoint, Replicate)
```


```{r loadSampleInfoB, echo=FALSE}
arrange(sampleinfo, Status, TimePoint, Replicate)
```

## Reading in the count data

* Salmon was used to quantify gene expression from raw reads.
* We need to read the data into R from the `quant.sf` files
* We use the `tximport` package to do this
* Salmon quantified expression at transcript level, we need to summarise this to gene level ...
* ... so we also need a table relating Transcript IDs to Gene IDs


```{r readSalmon, echo=FALSE, include=FALSE}
library(tximport)
salmon_files <- file.path("salmon", sampleinfo$SampleName, "quant.sf")
salmon_files <- set_names(salmon_files, sampleinfo$SampleName)
tx2gene <- read_tsv("references/tx2gene.tsv", show_col_types = FALSE)
txi <- tximport(salmon_files, type = "salmon", tx2gene = tx2gene)
```

```{r readSalmonF, eval=FALSE}
tx2gene <- read_tsv("references/tx2gene.tsv")
tx2gene
```

```{r readSalmonD, echo=FALSE}
tx2gene %>% print(n=6)
```

## Reading in the count data

```{r loadSalmon, eval=FALSE}
library(tximport)
txi <- tximport(salmon_files, type = "salmon", tx2gene = tx2gene)
str(txi)
```


```{r loadSalmonF, echo=FALSE}
str(txi)
```

## The counts matrix

```{r}
class(txi$counts)
typeof(txi$counts)
txi$counts[1:8, 1:5]
```


```{r rawCounts, echo=FALSE, include=FALSE}
rawCounts <- round(txi$counts, 0) 
keep <- rowSums(rawCounts) > 5
filtCounts <- rawCounts[keep,]
```

## Total counts per sample

* Ideally we look for > 20 million reads per sample (more if we are interested
in very lowly expressed genes)
* Samples with < 10 million reads may be problematic 

```{r, echo=FALSE, fig.align="center"}
rawCounts %>%
  as.data.frame() %>% 
  summarise(across(everything(), sum)) %>% 
  pivot_longer(names_to="SampleName", values_to="Counts", everything()) %>% 
  left_join(sampleinfo, by="SampleName") %>% 
  mutate(SampleGroup = str_c(Status, ".", TimePoint)) %>% 
  ggplot(aes(x=SampleName, y=Counts)) +
    geom_col(aes(fill = SampleGroup), colour="black") +
    geom_hline(yintercept = 20e6, colour="#33a02c", linetype=2) +
    geom_hline(yintercept = 10e6, colour="#e31a1c", linetype=2) +
    facet_wrap(vars(SampleGroup), nrow=1, scales = "free_x") +
    scale_colour_manual(values=c("#fdbf6f", "#ff7f00", "#a6cee3", "#1f78b4")) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    labs(x="") +
    guides(fill="none") +
    theme(plot.background = element_rect(fill='transparent', color=NA),
          panel.background = element_rect(fill='transparent', color=NA))

```

## Distribution of counts per gene

* There are a large number of genes with very low counts, we first removed these
before carrying on with data exploration.

```{r rawCounts2, eval=FALSE}
rawCounts <- round(txi$counts, 0) 
keep <- rowSums(rawCounts) > 5
filtCounts <- rawCounts[keep,]
```

## Distribution of counts per gene

* There are a large number of genes with very low counts, we first removed these
before carrying on with data exploration.

* We also need to tranform the data because:
  * There is a very large range in the raw counts:
  
```{r raw_summary2}
summary(filtCounts[,1:4])
```


## Distribution of counts per gene

* There are a large number of genes with very low counts, we first removed these
before carrying on with data exploration.

* We also need to tranform the data because:
  * There is a very large range in the raw counts:

```{r, echo=FALSE, fig.align="center", fig.height=4}
filtCounts %>%
  as.data.frame() %>% 
  rownames_to_column("GeneID") %>% 
  pivot_longer(names_to="SampleName", values_to="Counts", -GeneID) %>% 
  left_join(sampleinfo, by="SampleName") %>% 
  mutate(SampleGroup = str_c(Status, ".", TimePoint)) %>% 
  ggplot(aes(x=Counts)) +
    geom_density(aes(colour = SampleGroup)) +
    scale_colour_manual(values=c("#fdbf6f", "#ff7f00", "#a6cee3", "#1f78b4"))

```

## Distribution of counts per gene

* There are a large number of genes with very low counts, we first removed these
before carrying on with data exploration.

* We also need to tranform the data because:
  * There is a very large range in the raw counts
  * There is a relationship between expression and variance:
  
```{r raw_mean_vs_sda, echo=FALSE, fig.align="center", warning=FALSE, fig.width=5, fig.height=3.5}
tibble(MeanExp=rowMeans(filtCounts), 
       stdev=rowSds(filtCounts)) %>% 
  ggplot(aes(x= MeanExp, y=stdev)) +
    geom_point(fill="#A55532", colour="black", shape=21) +
    labs(x = "Mean Counts", y = "Standard Deviation") +
    lims(x = c(0,10000), y = c(0,5000))
```

## Data transformation

To avoid problems posed by raw counts, they can be transformed.
DESeq2 provides 2 transformation methods:

* VST : variance stabilizing transformation
* rlog : regularized log transformation

Both log2 transform the data and attempt to stabilise the variance with regard
to the mean expression of the gene. 

* rlog is slower but is more suitable when there is a very large range of counts
across samples (10x difference or more)
* VST is faster and less sensitive to large count outliers
* Either is suitable but generally use rlog on small dataset and VST on larger 
ones

## Data transformation

```{r vst_countsa, message=FALSE}
rlogCounts <- rlog(filtCounts)
```

```{r rlog_mean_vs_sd2, echo=FALSE, fig.align="center", warning=FALSE}
tibble(MeanExp=rowMeans(rlogCounts), 
       stdev=rowSds(rlogCounts)) %>% 
  ggplot(aes(x= MeanExp, y=stdev)) +
    geom_point(fill="#A55532", colour="black", shape=21) +
    labs(x = "Mean rlog(Counts)", y = "Standard Deviation")
```

## Distibution of transformed counts

```{r, echo=FALSE, fig.align="center"}
rlogCounts %>%
  as.data.frame() %>% 
  rownames_to_column("GeneID") %>% 
  pivot_longer(names_to="SampleName", values_to="Counts", -GeneID) %>% 
  left_join(sampleinfo, by="SampleName") %>% 
  mutate(SampleGroup = str_c(Status, ".", TimePoint)) %>% 
  ggplot(aes(x=Counts)) +
    geom_density(aes(colour = SampleGroup)) +
    scale_colour_manual(values=c("#fdbf6f", "#ff7f00", "#a6cee3", "#1f78b4")) +
    labs(x = "rlog(Counts)")

```

## Principal Component Analysis

* A principal component analysis (PCA) is an example of an unsupervised analysis 
* If the experiment has worked well we should find that:
    * replicate samples cluster closely
    * the greatest sources of variation should be between treatments/sample groups
* It is also an incredibly useful tool for checking for outliers and batch effects.
* We use the top most variable genes for the PCA (in this case I have used the top 500).

## Principal Component Analysis

```{r pcaPlot, echo=FALSE, fig.width=6.5, fig.height=5, fig.align="center"}
library(ggfortify)
rvs <- rowVars(rlogCounts)
hvgs <- order(rvs, decreasing = TRUE)[1:500]

pcDat <- prcomp(t(rlogCounts[hvgs,]))

autoplot(pcDat,
         data = sampleinfo, 
         fill="Status",
         shape="TimePoint",
         size=5)  +
  scale_shape_manual(values=c(21, 24)) +
  guides(fill = guide_legend(override.aes=list(shape=21)))
```

## Principal Component Analysis - Sample Swap

```{r pcaPlotss, echo=FALSE, fig.width=6.5, fig.height=5, fig.align="center"}
library(ggrepel)

autoplot(pcDat,
         data = sampleinfo, 
         fill="Status", 
         shape="TimePoint",
         size=5)+
    geom_text_repel(aes(x=PC1, y=PC2, label=SampleName), box.padding = 0.8)  +
  scale_shape_manual(values=c(21, 24)) +
  guides(fill = guide_legend(override.aes=list(shape=21)))
```

## Principal Component Analysis - Fixed

```{r correctSampleSheet}
sampleinfo <- mutate(sampleinfo, Status=case_when(
                                          SampleName=="SRR7657882" ~ "Uninfected",
                                          SampleName=="SRR7657873" ~ "Infected", 
                                          TRUE ~ Status))
```

```{r correctSampleSheetPCA, echo=FALSE, fig.width=5, fig.height=4, fig.align="center"}
autoplot(pcDat,
         data = sampleinfo, 
         fill="Status", 
         shape="TimePoint",
         size=5)  +
  scale_shape_manual(values=c(21, 24)) +
  guides(fill = guide_legend(override.aes=list(shape=21)))
```

## More exploratory plots

See the **Extended Materials** ...

<img src="images/ExtraPlots.png" class="centerimg" style="width: 60%;">